source/simulator/introspection.cc

```diff
@@ -32,166 +32,155 @@ namespace aspect
+    template <int dim>
+    std_cxx11::shared_ptr<FiniteElement<dim> >
+    new_FE_Q_or_DGP(const bool discontinuous,
+                    const unsigned int degree)
+    {
+      if (discontinuous)
+        return std_cxx11::shared_ptr<FiniteElement<dim> >(new FE_DGP<dim>(degree));
       else
-        fes.push_back(new FE_Q<dim>(parameters.composition_degree));
-      multiplicities.push_back(parameters.n_compositional_fields);
-      bes.compositional_fields = base_element++;
-
-      Assert(base_element == fes.size(), ExcInternalError());
-      Assert(base_element == multiplicities.size(), ExcInternalError());
-
-      return std_cxx11::make_tuple(bes, fes, multiplicities);
+        return std_cxx11::shared_ptr<FiniteElement<dim> >(new FE_Q<dim>(degree));
     }
 
-    /**
-     * Construct mapping from component to block indices.
-     */
     template <int dim>
-    std::vector<unsigned int>
-    setup_component_to_blocks (const typename Introspection<dim>::ComponentIndices &component_indices,
-                               const typename Introspection<dim>::BlockIndices &block_indices,
-                               const unsigned int n_components)
+    std_cxx11::shared_ptr<FiniteElement<dim> >
+    new_FE_Q_or_DGQ(const bool discontinuous,
+                    const unsigned int degree)
     {
-      std::vector<unsigned int> components_to_blocks;
-      const unsigned int n_compositional_fields = component_indices.compositional_fields.size();
-
-      components_to_blocks.resize(n_components, dealii::numbers::invalid_unsigned_int);
-      for (unsigned int d=0; d<dim; ++d)
-        components_to_blocks[component_indices.velocities[d]] = block_indices.velocities;
-      components_to_blocks[component_indices.pressure] = block_indices.pressure;
-      components_to_blocks[component_indices.temperature] = block_indices.temperature;
-      for (unsigned int c=0; c<n_compositional_fields; ++c)
-        components_to_blocks[component_indices.compositional_fields[c]] = block_indices.compositional_fields[c];
-
-#ifdef DEBUG
-      // check we assigned all components
-      for (unsigned int c=0; c<n_components; ++c)
-        Assert(components_to_blocks[c]!=dealii::numbers::invalid_unsigned_int, ExcInternalError());
-#endif
-
-      return components_to_blocks;
+      if (discontinuous)
+        return std_cxx11::shared_ptr<FiniteElement<dim> >(new FE_DGQ<dim>(degree));
+      else
+        return std_cxx11::shared_ptr<FiniteElement<dim> >(new FE_Q<dim>(degree));
     }
   }
 
 
 
+
   template <int dim>
-  Introspection<dim>::Introspection(const Parameters<dim> &parameters)
+  std::vector<VariableDeclaration<dim> >
+  construct_default_variables (const Parameters<dim> &parameters)
+  {
+    std::vector<VariableDeclaration<dim> > variables;
+    const bool use_direct_solver = parameters.use_direct_stokes_solver;
+
+
+    const unsigned int n_velocity_blocks = use_direct_solver?0:1;
```

Comment:
it took me a long time to see that `use_direct_solver` is used in this line because I though it was still used further down. Just put lines 119 and 122 next to each other to make the connection clearer.